lib/Basic/RedirectedProcess.cpp

```diff
@@ -0,0 +1,217 @@
+  } else {
+    posix_spawn_file_actions_adddup2(&fileActions, 1, 1);
+    posix_spawn_file_actions_adddup2(&fileActions, 2, 2);
+  }
+
+  // Spawn the command.
+  pid_t pid;
+  {
+    // We need to hold the spawn processes lock when we spawn, to ensure that
+    // we don't create a process in between when we are cancelled.
+    std::lock_guard<std::mutex> guard(spawnedProcessesMutex);
+
+    if (posix_spawn(&pid, path, /*file_actions=*/&fileActions,
+                    /*attrp=*/&attributes, args, envp) != 0) {
+      return false;
+    }
+  }
+
+  posix_spawn_file_actions_destroy(&fileActions);
+  posix_spawnattr_destroy(&attributes);
+
+  innerProcessInfo->processId = pid;
+
+  return true;
+}
+
+bool RedirectedProcess::readPipe(llvm::SmallString<1024> &output) {
+  if (!shouldCaptureOutput) {
+    return true;
+  }
+
+  // Close the write end of the output pipe.
+  ::close(innerProcessInfo->writeHandle);
+
+  // Read all the data from the output pipe.
+  while (true) {
+    char buf[4096];
+    ssize_t numBytes = ::read(innerProcessInfo->readHandle, buf, sizeof(buf));
+    if (numBytes < 0) {
+      return false;
+    }
+
+    if (numBytes == 0)
+      break;
+
+    output.insert(output.end(), &buf[0], &buf[numBytes]);
+  }
+
+  // Close the read end of the pipe.
+  ::close(innerProcessInfo->readHandle);
+
+  return true;
+}
+
+bool RedirectedProcess::waitForCompletion(int *exitStatus) {
+  int result = waitpid(innerProcessInfo->processId, exitStatus, 0);
+  while (result == -1 && errno == EINTR)
+    result = waitpid(innerProcessInfo->processId, exitStatus, 0);
+
+  return result != -1;
+}
+
+bool RedirectedProcess::kill(int signal) {
+  int result = ::kill(-(innerProcessInfo->processId), signal);
+  return result == 0;
+}
+
+bool RedirectedProcess::operator==(const RedirectedProcess &rhs) const {
+  return innerProcessInfo->processId == rhs.innerProcessInfo->processId &&
+         innerProcessInfo->readHandle == rhs.innerProcessInfo->readHandle &&
+         innerProcessInfo->writeHandle == rhs.innerProcessInfo->writeHandle;
+}
+
+size_t RedirectedProcess::hash() const {
+  return innerProcessInfo->processId;
+}
+
+RedirectedProcess::~RedirectedProcess() {
+  // TODO: this causes tens of errors at runtime running the tests:
+  // "*** Error in `/mnt/c/Users/hughb/Documents/GitHub/swift-linux/build/bin/llbuild': double free or corruption (fasttop): 0x00007f1450000940 ***"
+  // Surely we allocated innerProcessInfo (this->innerProcessInfo = new ProcessInfo();), so have
+  // to delete it? This may need investigation to make sure we don't leak memory.
+  // Or I'm probably wrong, and should delete this destructor.
+
+  //delete innerProcessInfo;
+}
+
+
+int RedirectedProcess::sigkill() {
+  return SIGKILL;
+}
+
+bool RedirectedProcess::isProcessCancelledStatus(int status) {
+  return WIFSIGNALED(status) &&
+    (WTERMSIG(status) == SIGINT || WTERMSIG(status) == SIGKILL);
```

Comment:
API simplification: rather than have the client pass back in the status which we previously handed them, why not return a more rich status object, or alternately just store all the status here and (a) assert if it isn't present and (b) avoid forcing the client to pass it back in?